File: expense_tracker.go
// expense_tracker.go - Трекер расходов на Go (CLI)
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

type Expense struct {
	ID          int     `json:"id"`
	Category    string  `json:"category"`
	Amount      float64 `json:"amount"`
	Date        string  `json:"date"`
	Description string  `json:"description"`
}

type Tracker struct {
	Expenses   []Expense       `json:"expenses"`
	NextID     int             `json:"next_id"`
	Categories map[string]bool `json:"-"`
}

const dataFile = "expenses.json"

func (t *Tracker) load() {
	file, err := os.ReadFile(dataFile)
	if err != nil {
		t.Expenses = []Expense{}
		t.NextID = 1
		t.Categories = make(map[string]bool)
		return
	}
	err = json.Unmarshal(file, t)
	if err != nil {
		t.Expenses = []Expense{}
		t.NextID = 1
	}
	t.Categories = make(map[string]bool)
	for _, e := range t.Expenses {
		t.Categories[e.Category] = true
	}
}

func (t *Tracker) save() {
	data, _ := json.MarshalIndent(t, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func (t *Tracker) add(category string, amount float64, date string, description string) Expense {
	if date == "" {
		date = time.Now().Format("2006-01-02")
	}
	e := Expense{
		ID:          t.NextID,
		Category:    category,
		Amount:      amount,
		Date:        date,
		Description: description,
	}
	t.Expenses = append(t.Expenses, e)
	t.NextID++
	t.Categories[category] = true
	t.save()
	return e
}

func (t *Tracker) edit(id int, category *string, amount *float64, date *string, description *string) bool {
	for i, e := range t.Expenses {
		if e.ID == id {
			if category != nil {
				t.Expenses[i].Category = *category
				t.Categories[*category] = true
			}
			if amount != nil {
				t.Expenses[i].Amount = *amount
			}
			if date != nil {
				t.Expenses[i].Date = *date
			}
			if description != nil {
				t.Expenses[i].Description = *description
			}
			t.save()
			return true
		}
	}
	return false
}

func (t *Tracker) delete(id int) bool {
	for i, e := range t.Expenses {
		if e.ID == id {
			t.Expenses = append(t.Expenses[:i], t.Expenses[i+1:]...)
			t.save()
			return true
		}
	}
	return false
}

func (t *Tracker) getExpenses(category *string, dateFrom, dateTo *string, minAmount, maxAmount *float64) []Expense {
	result := t.Expenses
	if category != nil {
		result = filterByCategory(result, *category)
	}
	if dateFrom != nil {
		result = filterByDateFrom(result, *dateFrom)
	}
	if dateTo != nil {
		result = filterByDateTo(result, *dateTo)
	}
	if minAmount != nil {
		result = filterByMinAmount(result, *minAmount)
	}
	if maxAmount != nil {
		result = filterByMaxAmount(result, *maxAmount)
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].Date < result[j].Date
	})
	return result
}

func filterByCategory(expenses []Expense, cat string) []Expense {
	var res []Expense
	for _, e := range expenses {
		if e.Category == cat {
			res = append(res, e)
		}
	}
	return res
}

func filterByDateFrom(expenses []Expense, from string) []Expense {
	var res []Expense
	for _, e := range expenses {
		if e.Date >= from {
			res = append(res, e)
		}
	}
	return res
}

func filterByDateTo(expenses []Expense, to string) []Expense {
	var res []Expense
	for _, e := range expenses {
		if e.Date <= to {
			res = append(res, e)
		}
	}
	return res
}

func filterByMinAmount(expenses []Expense, min float64) []Expense {
	var res []Expense
	for _, e := range expenses {
		if e.Amount >= min {
			res = append(res, e)
		}
	}
	return res
}

func filterByMaxAmount(expenses []Expense, max float64) []Expense {
	var res []Expense
	for _, e := range expenses {
		if e.Amount <= max {
			res = append(res, e)
		}
	}
	return res
}

func (t *Tracker) getStatistics(category *string, dateFrom, dateTo *string) (total float64, count int, byCategory map[string]float64, avg float64) {
	expenses := t.getExpenses(category, dateFrom, dateTo, nil, nil)
	total = 0
	byCategory = make(map[string]float64)
	for _, e := range expenses {
		total += e.Amount
		byCategory[e.Category] += e.Amount
	}
	count = len(expenses)
	if count > 0 {
		avg = total / float64(count)
	}
	return
}

func (t *Tracker) listCategories() []string {
	var cats []string
	for c := range t.Categories {
		cats = append(cats, c)
	}
	sort.Strings(cats)
	return cats
}

func main() {
	var (
		cmd         string
		category    string
		amount      float64
		date        string
		description string
		id          int
		from        string
		to          string
		minAmt      float64
		maxAmt      float64
	)
	flag.StringVar(&cmd, "cmd", "", "Команда: add, list, stats, delete, edit, categories")
	flag.StringVar(&category, "category", "", "Категория")
	flag.Float64Var(&amount, "amount", 0, "Сумма")
	flag.StringVar(&date, "date", "", "Дата")
	flag.StringVar(&description, "description", "", "Описание")
	flag.IntVar(&id, "id", 0, "ID записи")
	flag.StringVar(&from, "from", "", "Дата от")
	flag.StringVar(&to, "to", "", "Дата до")
	flag.Float64Var(&minAmt, "min", 0, "Мин. сумма")
	flag.Float64Var(&maxAmt, "max", 0, "Макс. сумма")
	flag.Parse()

	tracker := &Tracker{}
	tracker.load()

	switch cmd {
	case "add":
		if category == "" || amount <= 0 {
			fmt.Println("Необходимо указать --category и --amount")
			return
		}
		if date == "" {
			date = time.Now().Format("2006-01-02")
		}
		e := tracker.add(category, amount, date, description)
		fmt.Printf("✅ Добавлен расход #%d: %.2f %s (%s)\n", e.ID, e.Amount, e.Category, e.Date)
	case "list":
		expenses := tracker.getExpenses(&category, &from, &to, &minAmt, &maxAmt)
		if len(expenses) == 0 {
			fmt.Println("Нет записей.")
		} else {
			total := 0.0
			fmt.Printf("%-4s %-12s %-15s %-10s %s\n", "ID", "Дата", "Категория", "Сумма", "Описание")
			for _, e := range expenses {
				fmt.Printf("%-4d %-12s %-15s %-10.2f %s\n", e.ID, e.Date, e.Category, e.Amount, e.Description)
				total += e.Amount
			}
			fmt.Printf("\nИтого: %.2f\n", total)
		}
	case "stats":
		var catPtr *string
		if category != "" {
			catPtr = &category
		}
		var fromPtr, toPtr *string
		if from != "" {
			fromPtr = &from
		}
		if to != "" {
			toPtr = &to
		}
		total, count, byCategory, avg := tracker.getStatistics(catPtr, fromPtr, toPtr)
		fmt.Printf("📊 Статистика\n")
		fmt.Printf("Всего записей: %d\n", count)
		fmt.Printf("Общая сумма: %.2f\n", total)
		fmt.Printf("Средний расход: %.2f\n", avg)
		if len(byCategory) > 0 {
			fmt.Println("По категориям:")
			type kv struct {
				Key   string
				Value float64
			}
			var sorted []kv
			for k, v := range byCategory {
				sorted = append(sorted, kv{k, v})
			}
			sort.Slice(sorted, func(i, j int) bool {
				return sorted[i].Value > sorted[j].Value
			})
			for _, kv := range sorted {
				percent := (kv.Value / total) * 100
				if total == 0 {
					percent = 0
				}
				fmt.Printf("  %s: %.2f (%.1f%%)\n", kv.Key, kv.Value, percent)
			}
		}
	case "delete":
		if id == 0 {
			fmt.Println("Укажите --id")
			return
		}
		if tracker.delete(id) {
			fmt.Printf("✅ Расход #%d удалён\n", id)
		} else {
			fmt.Printf("❌ Расход #%d не найден\n", id)
		}
	case "edit":
		if id == 0 {
			fmt.Println("Укажите --id")
			return
		}
		var catPtr *string
		var amtPtr *float64
		var datePtr *string
		var descPtr *string
		if category != "" {
			catPtr = &category
		}
		if amount > 0 {
			amtPtr = &amount
		}
		if date != "" {
			datePtr = &date
		}
		if description != "" {
			descPtr = &description
		}
		if tracker.edit(id, catPtr, amtPtr, datePtr, descPtr) {
			fmt.Printf("✅ Расход #%d обновлён\n", id)
		} else {
			fmt.Printf("❌ Расход #%d не найден\n", id)
		}
	case "categories":
		cats := tracker.listCategories()
		if len(cats) > 0 {
			fmt.Println("Категории:", strings.Join(cats, ", "))
		} else {
			fmt.Println("Нет категорий")
		}
	default:
		// interactive mode
		interactiveMode(tracker)
	}
}

func interactiveMode(tracker *Tracker) {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Println("\n💰 Трекер расходов (интерактивный)")
		fmt.Println("1. Добавить расход")
		fmt.Println("2. Показать расходы (с фильтром)")
		fmt.Println("3. Статистика")
		fmt.Println("4. Редактировать")
		fmt.Println("5. Удалить")
		fmt.Println("6. Список категорий")
		fmt.Println("0. Выход")
		fmt.Print("Выберите действие: ")
		scanner.Scan()
		choice := scanner.Text()
		switch choice {
		case "0":
			return
		case "1":
			fmt.Print("Категория: ")
			scanner.Scan()
			cat := scanner.Text()
			if cat == "" {
				fmt.Println("Категория обязательна")
				continue
			}
			fmt.Print("Сумма: ")
			scanner.Scan()
			amtStr := scanner.Text()
			amt, err := strconv.ParseFloat(amtStr, 64)
			if err != nil || amt <= 0 {
				fmt.Println("Неверная сумма")
				continue
			}
			fmt.Print("Дата (ГГГГ-ММ-ДД, Enter сегодня): ")
			scanner.Scan()
			date := scanner.Text()
			if date == "" {
				date = time.Now().Format("2006-01-02")
			}
			fmt.Print("Описание (необязательно): ")
			scanner.Scan()
			desc := scanner.Text()
			e := tracker.add(cat, amt, date, desc)
			fmt.Printf("✅ Добавлен расход #%d\n", e.ID)
		case "2":
			fmt.Print("Категория (Enter пропустить): ")
			scanner.Scan()
			cat := scanner.Text()
			var catPtr *string
			if cat != "" {
				catPtr = &cat
			}
			fmt.Print("Дата от (Enter пропустить): ")
			scanner.Scan()
			from := scanner.Text()
			var fromPtr *string
			if from != "" {
				fromPtr = &from
			}
			fmt.Print("Дата до (Enter пропустить): ")
			scanner.Scan()
			to := scanner.Text()
			var toPtr *string
			if to != "" {
				toPtr = &to
			}
			expenses := tracker.getExpenses(catPtr, fromPtr, toPtr, nil, nil)
			if len(expenses) == 0 {
				fmt.Println("Нет записей.")
			} else {
				total := 0.0
				fmt.Printf("%-4s %-12s %-15s %-10s %s\n", "ID", "Дата", "Категория", "Сумма", "Описание")
				for _, e := range expenses {
					fmt.Printf("%-4d %-12s %-15s %-10.2f %s\n", e.ID, e.Date, e.Category, e.Amount, e.Description)
					total += e.Amount
				}
				fmt.Printf("\nИтого: %.2f\n", total)
			}
		case "3":
			fmt.Print("Категория (Enter пропустить): ")
			scanner.Scan()
			cat := scanner.Text()
			var catPtr *string
			if cat != "" {
				catPtr = &cat
			}
			fmt.Print("Дата от (Enter пропустить): ")
			scanner.Scan()
			from := scanner.Text()
			var fromPtr *string
			if from != "" {
				fromPtr = &from
			}
			fmt.Print("Дата до (Enter пропустить): ")
			scanner.Scan()
			to := scanner.Text()
			var toPtr *string
			if to != "" {
				toPtr = &to
			}
			total, count, byCategory, avg := tracker.getStatistics(catPtr, fromPtr, toPtr)
			fmt.Printf("📊 Статистика\n")
			fmt.Printf("Всего записей: %d\n", count)
			fmt.Printf("Общая сумма: %.2f\n", total)
			fmt.Printf("Средний расход: %.2f\n", avg)
			if len(byCategory) > 0 {
				fmt.Println("По категориям:")
				type kv struct {
					Key   string
					Value float64
				}
				var sorted []kv
				for k, v := range byCategory {
					sorted = append(sorted, kv{k, v})
				}
				sort.Slice(sorted, func(i, j int) bool {
					return sorted[i].Value > sorted[j].Value
				})
				for _, kv := range sorted {
					percent := (kv.Value / total) * 100
					if total == 0 {
						percent = 0
					}
					fmt.Printf("  %s: %.2f (%.1f%%)\n", kv.Key, kv.Value, percent)
				}
			}
		case "4":
			fmt.Print("ID расхода: ")
			scanner.Scan()
			idStr := scanner.Text()
			id, err := strconv.Atoi(idStr)
			if err != nil {
				fmt.Println("Неверный ID")
				continue
			}
			var exp *Expense
			for _, e := range tracker.Expenses {
				if e.ID == id {
					exp = &e
					break
				}
			}
			if exp == nil {
				fmt.Println("Запись не найдена")
				continue
			}
			fmt.Println("Оставьте пустым, чтобы не менять.")
			fmt.Printf("Категория (%s): ", exp.Category)
			scanner.Scan()
			newCat := scanner.Text()
			fmt.Printf("Сумма (%.2f): ", exp.Amount)
			scanner.Scan()
			newAmtStr := scanner.Text()
			var newAmt *float64
			if newAmtStr != "" {
				if a, err := strconv.ParseFloat(newAmtStr, 64); err == nil {
					newAmt = &a
				}
			}
			fmt.Printf("Дата (%s): ", exp.Date)
			scanner.Scan()
			newDate := scanner.Text()
			var newDatePtr *string
			if newDate != "" {
				newDatePtr = &newDate
			}
			fmt.Printf("Описание (%s): ", exp.Description)
			scanner.Scan()
			newDesc := scanner.Text()
			var newDescPtr *string
			if newDesc != "" {
				newDescPtr = &newDesc
			}
			var catPtr *string
			if newCat != "" {
				catPtr = &newCat
			}
			if tracker.edit(id, catPtr, newAmt, newDatePtr, newDescPtr) {
				fmt.Println("✅ Обновлено")
			} else {
				fmt.Println("❌ Ошибка обновления")
			}
		case "5":
			fmt.Print("ID для удаления: ")
			scanner.Scan()
			idStr := scanner.Text()
			id, err := strconv.Atoi(idStr)
			if err != nil {
				fmt.Println("Неверный ID")
				continue
			}
			if tracker.delete(id) {
				fmt.Println("✅ Удалено")
			} else {
				fmt.Println("❌ Не найдено")
			}
		case "6":
			cats := tracker.listCategories()
			if len(cats) > 0 {
				fmt.Println("Категории:", strings.Join(cats, ", "))
			} else {
				fmt.Println("Нет категорий")
			}
		default:
			fmt.Println("Неверный выбор")
		}
	}
}

File: expense_tracker.rs
// expense_tracker.rs - Трекер расходов на Rust (CLI)
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::Path;
use std::str::FromStr;
use chrono::NaiveDate;

#[derive(Serialize, Deserialize, Clone)]
struct Expense {
    id: u32,
    category: String,
    amount: f64,
    date: String,
    description: String,
}

impl Expense {
    fn new(id: u32, category: &str, amount: f64, date: &str, description: &str) -> Self {
        Expense {
            id,
            category: category.to_string(),
            amount,
            date: date.to_string(),
            description: description.to_string(),
        }
    }
}

struct Tracker {
    expenses: Vec<Expense>,
    next_id: u32,
    categories: std::collections::HashSet<String>,
}

impl Tracker {
    fn new() -> Self {
        let mut t = Tracker {
            expenses: Vec::new(),
            next_id: 1,
            categories: std::collections::HashSet::new(),
        };
        t.load();
        t
    }

    fn load(&mut self) {
        let path = "expenses.json";
        if Path::new(path).exists() {
            if let Ok(data) = fs::read_to_string(path) {
                if let Ok(parsed) = serde_json::from_str::<Vec<Expense>>(&data) {
                    self.expenses = parsed;
                    self.next_id = self.expenses.iter().map(|e| e.id).max().unwrap_or(0) + 1;
                    self.categories = self.expenses.iter().map(|e| e.category.clone()).collect();
                    return;
                }
            }
        }
        self.expenses = Vec::new();
        self.next_id = 1;
        self.categories = std::collections::HashSet::new();
    }

    fn save(&self) {
        let data = serde_json::to_string_pretty(&self.expenses).unwrap();
        fs::write("expenses.json", data).unwrap();
    }

    fn add(&mut self, category: &str, amount: f64, date: &str, description: &str) -> Expense {
        let exp = Expense::new(self.next_id, category, amount, date, description);
        self.expenses.push(exp.clone());
        self.next_id += 1;
        self.categories.insert(category.to_string());
        self.save();
        exp
    }

    fn edit(&mut self, id: u32, category: Option<&str>, amount: Option<f64>, date: Option<&str>, description: Option<&str>) -> bool {
        for exp in &mut self.expenses {
            if exp.id == id {
                if let Some(c) = category {
                    exp.category = c.to_string();
                    self.categories.insert(c.to_string());
                }
                if let Some(a) = amount {
                    exp.amount = a;
                }
                if let Some(d) = date {
                    exp.date = d.to_string();
                }
                if let Some(d) = description {
                    exp.description = d.to_string();
                }
                self.save();
                return true;
            }
        }
        false
    }

    fn delete(&mut self, id: u32) -> bool {
        let len = self.expenses.len();
        self.expenses.retain(|e| e.id != id);
        if self.expenses.len() < len {
            self.save();
            return true;
        }
        false
    }

    fn get_expenses(&self, category: Option<&str>, date_from: Option<&str>, date_to: Option<&str>,
                   min_amount: Option<f64>, max_amount: Option<f64>) -> Vec<Expense> {
        let mut result = self.expenses.clone();
        if let Some(cat) = category {
            result.retain(|e| e.category == cat);
        }
        if let Some(from) = date_from {
            result.retain(|e| e.date >= from);
        }
        if let Some(to) = date_to {
            result.retain(|e| e.date <= to);
        }
        if let Some(min) = min_amount {
            result.retain(|e| e.amount >= min);
        }
        if let Some(max) = max_amount {
            result.retain(|e| e.amount <= max);
        }
        result.sort_by(|a, b| a.date.cmp(&b.date));
        result
    }

    fn get_statistics(&self, category: Option<&str>, date_from: Option<&str>, date_to: Option<&str>) -> (f64, usize, HashMap<String, f64>, f64) {
        let expenses = self.get_expenses(category, date_from, date_to, None, None);
        let total = expenses.iter().map(|e| e.amount).sum();
        let count = expenses.len();
        let mut by_category = HashMap::new();
        for e in &expenses {
            *by_category.entry(e.category.clone()).or_insert(0.0) += e.amount;
        }
        let avg = if count > 0 { total / count as f64 } else { 0.0 };
        (total, count, by_category, avg)
    }

    fn list_categories(&self) -> Vec<String> {
        let mut v: Vec<_> = self.categories.iter().cloned().collect();
        v.sort();
        v
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        interactive_mode();
        return;
    }
    let mut tracker = Tracker::new();
    match args[1].as_str() {
        "add" => {
            let mut category = String::new();
            let mut amount = 0.0;
            let mut date = String::new();
            let mut description = String::new();
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--category" => { category = args[i+1].clone(); i += 2; }
                    "--amount" => { amount = args[i+1].parse().unwrap_or(0.0); i += 2; }
                    "--date" => { date = args[i+1].clone(); i += 2; }
                    "--description" => { description = args[i+1].clone(); i += 2; }
                    _ => { i += 1; }
                }
            }
            if category.is_empty() || amount <= 0.0 {
                println!("Укажите --category и --amount");
                return;
            }
            if date.is_empty() {
                date = chrono::Local::now().format("%Y-%m-%d").to_string();
            }
            let exp = tracker.add(&category, amount, &date, &description);
            println!("✅ Добавлен расход #{}: {:.2} {} ({})", exp.id, exp.amount, exp.category, exp.date);
        }
        "list" => {
            let mut category = None;
            let mut date_from = None;
            let mut date_to = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--category" => { category = Some(args[i+1].clone()); i += 2; }
                    "--from" => { date_from = Some(args[i+1].clone()); i += 2; }
                    "--to" => { date_to = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            let expenses = tracker.get_expenses(category.as_deref(), date_from.as_deref(), date_to.as_deref(), None, None);
            if expenses.is_empty() {
                println!("Нет записей.");
            } else {
                let total: f64 = expenses.iter().map(|e| e.amount).sum();
                println!("{:<4} {:<12} {:<15} {:<10} {}", "ID", "Дата", "Категория", "Сумма", "Описание");
                for e in &expenses {
                    println!("{:<4} {:<12} {:<15} {:<10.2} {}", e.id, e.date, e.category, e.amount, e.description);
                }
                println!("\nИтого: {:.2}", total);
            }
        }
        "stats" => {
            let mut category = None;
            let mut date_from = None;
            let mut date_to = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--category" => { category = Some(args[i+1].clone()); i += 2; }
                    "--from" => { date_from = Some(args[i+1].clone()); i += 2; }
                    "--to" => { date_to = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            let (total, count, by_category, avg) = tracker.get_statistics(category.as_deref(), date_from.as_deref(), date_to.as_deref());
            println!("📊 Статистика");
            println!("Всего записей: {}", count);
            println!("Общая сумма: {:.2}", total);
            println!("Средний расход: {:.2}", avg);
            if !by_category.is_empty() {
                println!("По категориям:");
                let mut sorted: Vec<_> = by_category.into_iter().collect();
                sorted.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
                for (cat, amount) in sorted {
                    let percent = if total > 0.0 { (amount / total) * 100.0 } else { 0.0 };
                    println!("  {}: {:.2} ({:.1}%)", cat, amount, percent);
                }
            }
        }
        "delete" => {
            let mut id = 0;
            let mut i = 2;
            while i < args.len() {
                if args[i] == "--id" {
                    id = args[i+1].parse().unwrap_or(0);
                    break;
                }
                i += 1;
            }
            if id == 0 {
                println!("Укажите --id");
                return;
            }
            if tracker.delete(id) {
                println!("✅ Расход #{} удалён", id);
            } else {
                println!("❌ Расход #{} не найден", id);
            }
        }
        "edit" => {
            let mut id = 0;
            let mut category = None;
            let mut amount = None;
            let mut date = None;
            let mut description = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--id" => { id = args[i+1].parse().unwrap_or(0); i += 2; }
                    "--category" => { category = Some(args[i+1].clone()); i += 2; }
                    "--amount" => { amount = Some(args[i+1].parse().unwrap_or(0.0)); i += 2; }
                    "--date" => { date = Some(args[i+1].clone()); i += 2; }
                    "--description" => { description = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            if id == 0 {
                println!("Укажите --id");
                return;
            }
            if tracker.edit(id, category.as_deref(), amount, date.as_deref(), description.as_deref()) {
                println!("✅ Расход #{} обновлён", id);
            } else {
                println!("❌ Расход #{} не найден", id);
            }
        }
        "categories" => {
            let cats = tracker.list_categories();
            if cats.is_empty() {
                println!("Нет категорий");
            } else {
                println!("Категории: {}", cats.join(", "));
            }
        }
        _ => {
            interactive_mode();
        }
    }
}

fn interactive_mode() {
    let mut tracker = Tracker::new();
    let stdin = io::stdin();
    let mut stdout = io::stdout();
    loop {
        println!("\n💰 Трекер расходов (интерактивный)");
        println!("1. Добавить расход");
        println!("2. Показать расходы (с фильтром)");
        println!("3. Статистика");
        println!("4. Редактировать");
        println!("5. Удалить");
        println!("6. Список категорий");
        println!("0. Выход");
        print!("Выберите действие: ");
        stdout.flush().unwrap();
        let mut choice = String::new();
        stdin.read_line(&mut choice).unwrap();
        match choice.trim() {
            "0" => break,
            "1" => {
                print!("Категория: ");
                stdout.flush().unwrap();
                let mut cat = String::new();
                stdin.read_line(&mut cat).unwrap();
                let cat = cat.trim();
                if cat.is_empty() {
                    println!("Категория обязательна");
                    continue;
                }
                print!("Сумма: ");
                stdout.flush().unwrap();
                let mut amt = String::new();
                stdin.read_line(&mut amt).unwrap();
                let amount = match amt.trim().parse::<f64>() {
                    Ok(v) if v > 0.0 => v,
                    _ => { println!("Неверная сумма"); continue; }
                };
                print!("Дата (ГГГГ-ММ-ДД, Enter сегодня): ");
                stdout.flush().unwrap();
                let mut date = String::new();
                stdin.read_line(&mut date).unwrap();
                let date = if date.trim().is_empty() {
                    chrono::Local::now().format("%Y-%m-%d").to_string()
                } else {
                    date.trim().to_string()
                };
                print!("Описание (необязательно): ");
                stdout.flush().unwrap();
                let mut desc = String::new();
                stdin.read_line(&mut desc).unwrap();
                let exp = tracker.add(cat, amount, &date, desc.trim());
                println!("✅ Добавлен расход #{}", exp.id);
            }
            "2" => {
                print!("Категория (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut cat = String::new();
                stdin.read_line(&mut cat).unwrap();
                let cat = if cat.trim().is_empty() { None } else { Some(cat.trim()) };
                print!("Дата от (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut from = String::new();
                stdin.read_line(&mut from).unwrap();
                let from = if from.trim().is_empty() { None } else { Some(from.trim()) };
                print!("Дата до (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut to = String::new();
                stdin.read_line(&mut to).unwrap();
                let to = if to.trim().is_empty() { None } else { Some(to.trim()) };
                let expenses = tracker.get_expenses(cat, from, to, None, None);
                if expenses.is_empty() {
                    println!("Нет записей.");
                } else {
                    let total: f64 = expenses.iter().map(|e| e.amount).sum();
                    println!("{:<4} {:<12} {:<15} {:<10} {}", "ID", "Дата", "Категория", "Сумма", "Описание");
                    for e in &expenses {
                        println!("{:<4} {:<12} {:<15} {:<10.2} {}", e.id, e.date, e.category, e.amount, e.description);
                    }
                    println!("\nИтого: {:.2}", total);
                }
            }
            "3" => {
                print!("Категория (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut cat = String::new();
                stdin.read_line(&mut cat).unwrap();
                let cat = if cat.trim().is_empty() { None } else { Some(cat.trim()) };
                print!("Дата от (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut from = String::new();
                stdin.read_line(&mut from).unwrap();
                let from = if from.trim().is_empty() { None } else { Some(from.trim()) };
                print!("Дата до (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut to = String::new();
                stdin.read_line(&mut to).unwrap();
                let to = if to.trim().is_empty() { None } else { Some(to.trim()) };
                let (total, count, by_category, avg) = tracker.get_statistics(cat, from, to);
                println!("📊 Статистика");
                println!("Всего записей: {}", count);
                println!("Общая сумма: {:.2}", total);
                println!("Средний расход: {:.2}", avg);
                if !by_category.is_empty() {
                    println!("По категориям:");
                    let mut sorted: Vec<_> = by_category.into_iter().collect();
                    sorted.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
                    for (cat, amount) in sorted {
                        let percent = if total > 0.0 { (amount / total) * 100.0 } else { 0.0 };
                        println!("  {}: {:.2} ({:.1}%)", cat, amount, percent);
                    }
                }
            }
            "4" => {
                print!("ID расхода: ");
                stdout.flush().unwrap();
                let mut id_str = String::new();
                stdin.read_line(&mut id_str).unwrap();
                let id = match id_str.trim().parse::<u32>() {
                    Ok(v) => v,
                    _ => { println!("Неверный ID"); continue; }
                };
                let exp = tracker.expenses.iter().find(|e| e.id == id).cloned();
                let exp = match exp {
                    Some(e) => e,
                    None => { println!("Запись не найдена"); continue; }
                };
                println!("Оставьте пустым, чтобы не менять.");
                print!("Категория ({}): ", exp.category);
                stdout.flush().unwrap();
                let mut new_cat = String::new();
                stdin.read_line(&mut new_cat).unwrap();
                let new_cat = if new_cat.trim().is_empty() { None } else { Some(new_cat.trim()) };
                print!("Сумма ({:.2}): ", exp.amount);
                stdout.flush().unwrap();
                let mut new_amt = String::new();
                stdin.read_line(&mut new_amt).unwrap();
                let new_amt = if new_amt.trim().is_empty() { None } else {
                    match new_amt.trim().parse::<f64>() {
                        Ok(v) => Some(v),
                        _ => None
                    }
                };
                print!("Дата ({}): ", exp.date);
                stdout.flush().unwrap();
                let mut new_date = String::new();
                stdin.read_line(&mut new_date).unwrap();
                let new_date = if new_date.trim().is_empty() { None } else { Some(new_date.trim()) };
                print!("Описание ({}): ", exp.description);
                stdout.flush().unwrap();
                let mut new_desc = String::new();
                stdin.read_line(&mut new_desc).unwrap();
                let new_desc = if new_desc.trim().is_empty() { None } else { Some(new_desc.trim()) };
                if tracker.edit(id, new_cat, new_amt, new_date, new_desc) {
                    println!("✅ Обновлено");
                } else {
                    println!("❌ Ошибка обновления");
                }
            }
            "5" => {
                print!("ID для удаления: ");
                stdout.flush().unwrap();
                let mut id_str = String::new();
                stdin.read_line(&mut id_str).unwrap();
                let id = match id_str.trim().parse::<u32>() {
                    Ok(v) => v,
                    _ => { println!("Неверный ID"); continue; }
                };
                if tracker.delete(id) {
                    println!("✅ Удалено");
                } else {
                    println!("❌ Не найдено");
                }
            }
            "6" => {
                let cats = tracker.list_categories();
                if cats.is_empty() {
                    println!("Нет категорий");
                } else {
                    println!("Категории: {}", cats.join(", "));
                }
            }
            _ => println!("Неверный выбор"),
        }
    }
}
